core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -52,50 +65,100 @@
             Log.error(Log.JEEVES, "Error occurred within a transaction", e);
             if (exception[0] == null) {
                 exception[0] = e;
             }
-            rolledBack = true;
-            doRollback(context, transactionManager, transaction);
+
+            try {
+                doRollback(context, transactionManager, status);
+            }
+            finally {
+                isRolledBack = true;
+            }
         } finally {
             try {
-                if (readOnly) {
-                    doRollback(context, transactionManager, transaction);
-                } else if (!rolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
-                    doCommit(context, transactionManager, transaction);
+                if (readOnly && !isRolledBack) {
+                    try {
+                        doRollback(context, transactionManager, status);
+                    }
+                    finally {
+                        isRolledBack = true;
+                    }
+                } else if (!isRolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
+                    try {
+                        doCommit(context, transactionManager, status);
+                    }
+                    finally {
+                        isCommitted = true;
+                    }
                 }
             } catch (TransactionSystemException e) {
                 if (!(e.getOriginalException() instanceof RollbackException)) {
                     Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
-                    doRollback(context, transactionManager, transaction);
+                    if (!isRolledBack) {
+                        try {
+                            doRollback(context, transactionManager, status);
+                        }
+                        finally {
+                            isRolledBack = true;
+                        }
+                    }
                 } else {
                     Log.debug(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
+                    if (!isRolledBack) {
+                        try {
+                            doRollback(context, transactionManager, status);
+                        }
+                        finally {
+                            isRolledBack = true;
+                        }
+                    }
                 }
             } catch (Throwable t) {
                 Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", t);
-                doRollback(context, transactionManager, transaction);
+                if (!isRolledBack) {
+                    try {
+                        doRollback(context, transactionManager, status);
+                    }
+                    finally {
+                        isRolledBack = true;
+                    }
+                }
```

Comment:
This represents a small behaviour change, avoiding calling doRollback twice - if a failure occurred during the course of calling `doRollback` above.